trl/vllm_utils.py

```diff
@@ -0,0 +1,256 @@
+
+    ```python
+    >>> from trl import VLLMServer
+    >>> server = VLLMServer(model_name="Qwen/Qwen2.5-7B-Instruct")
+    >>> server.run()
+    ```
+
+    Use the server to generate completions:
+    ```shell
+    $ curl -X POST "http://0.0.0.0:5000/generate" -H "Content-Type: application/json" -d '{"prompts": ["The closest planet to the Sun is"]}'
+    [" ____\nA. Earth\nB. Venus\nC. Mercury\nD."]
+    ```
+    """
+
+    def __init__(self, model_name=None, url: str = "http://0.0.0.0:5000"):
+        if not is_flask_available():
+            raise ImportError("vLLM server requires Flask. Please install it with `pip install flask`.")
+
+        if not is_vllm_available():
+            raise ImportError("vLLM server requires the `vllm` package. Please install it with `pip install vllm`.")
+
+        match = re.match(r"http://(.*):(\d+)", url)
+        if match is None:
+            raise ValueError(f"Invalid URL format: {url}")
+
+        self.host, self.port = match.groups()
+        self.app = Flask(__name__)  # Initialize Flask app
+
+        self._add_routes()  # Add routes to the app
+        if model_name is not None:
+            self.llm = LLM(model=model_name)
+        else:
+            self.llm = None
+
+    def _add_routes(self):
+        """Add the Flask routes for the server."""
+
+        @self.app.route("/load", methods=["POST"])
+        def load():
+            try:
+                data = request.get_json()
+                del self.llm  # First delete the existing model, to avoid OOM errors
+                self.llm = LLM(model=data["model_name"])
+                self.app.logger.info(f"Model {data['model_name']} loaded.")
+                return jsonify({"status": "success", "message": "Model loaded."})
+
+            except Exception as e:
+                self.app.logger.error(f"Error loading model: {str(e)}")
+                return jsonify({"error": str(e)}), 400
+
+        @self.app.route("/generate", methods=["POST"])
+        def generate():
+            if self.llm is None:
+                return jsonify({"error": "No model loaded. Load a model using the /load endpoint."}), 400
+
+            try:
+                # Parse input JSON data
+                data = request.get_json()
+
+                # Get prompts
+                prompts = data["prompts"]
+
+                # Get sampling params (optional)
+                sampling_params = SamplingParams(**data["sampling_params"]) if "sampling_params" in data else None
+
+                # Determine return type (default to 'text')
+                return_type = data.get("return_type", "text")
+                if return_type not in {"text", "tokens"}:
+                    return jsonify({"error": f"Invalid return_type '{return_type}'. Must be 'text' or 'tokens'."}), 400
+
+                # Generate completions
+                if is_conversational({"prompt": prompts[0]}):
+                    outputs = self.llm.chat(prompts, sampling_params=sampling_params)
```

Comment:
For my understanding, this case will be passing `messages` right?

Reply:
yes, prompts are messages here